File: profiles/subscribe_and_unsubscribe.py
# this sample Python code subscribes and unsubscribes a profile from email and SMS marketing.
# see https://developers.klaviyo.com/en/reference/bulk_subscribe_profiles and https://developers.klaviyo.com/en/reference/bulk_unsubscribe_profiles for more details.

import os
from klaviyo_api import KlaviyoAPI

# Load your private API key from environment variables. Note that this API key needs to have write access to profiles, lists, and subscriptions.
private_api_key = os.environ.get("private_api_key")
klaviyo = KlaviyoAPI(private_api_key)

# Example 1: Subscribe a profile to email and SMS marketing
customer_email = 'CUSTOMER_EMAIL'  # replace with the email address
customer_phone_number = '+15555555555'  # replace with the phone number
list_id = 'YOUR_LIST_ID'  # replace with your list ID

subscribe_profile_request_body = {
    "data": {
        "type": "profile-subscription-bulk-create-job",
        "attributes": {
            "profiles": {
                "data": [
                    {
                        "type": "profile",
                        "attributes": {
                            "subscriptions": {
                                "email": {
                                    "marketing": {
                                        "consent": "SUBSCRIBED"
                                    }
                                },
                                "sms": {
                                    "marketing": {
                                        "consent": "SUBSCRIBED"
                                    },
                                    "transactional": {
                                        "consent": "SUBSCRIBED"
                                    }
                                }
                            },
                            "email": customer_email,
                            "phone_number": customer_phone_number
                        }
                    }
                ]
            },
            "historical_import": False
        },
        "relationships": {
            "list": {
                "data": {
                    "type": "list",
                    "id": list_id
                }
            }
        }
    }
}

response = klaviyo.Profiles.bulk_subscribe_profiles(subscribe_profile_request_body)

# Example 2: Subscribe one profile to email marketing, and one to SMS marketing and transactional SMS
customer_email_1 = 'CUSTOMER_EMAIL_1'  # replace with first customer's email
customer_email_2 = 'CUSTOMER_EMAIL_2'  # replace with second customer's email
customer_phone_2 = '+15555555556'      # replace with second customer's phone number
list_id_bulk = 'YOUR_LIST_ID'          # replace with your list ID

subscribe_profiles_request_body = {
    "data": {
        "type": "profile-subscription-bulk-create-job",
        "attributes": {
            "profiles": {
                "data": [
                    {
                        "type": "profile",
                        "attributes": {
                            "subscriptions": {
                                "email": {
                                    "marketing": {
                                        "consent": "SUBSCRIBED"
                                    }
                                }
                            },
                            "email": customer_email_1
                        }
                    },
                    {
                        "type": "profile",
                        "attributes": {
                            "subscriptions": {
                                "sms": {
                                    "marketing": {
                                        "consent": "SUBSCRIBED"
                                    },
                                    "transactional": {
                                        "consent": "SUBSCRIBED"
                                    }
                                }
                            },
                            "email": customer_email_2,
                            "phone_number": customer_phone_2
                        }
                    }
                ]
            },
            "historical_import": False
        },
        "relationships": {
            "list": {
                "data": {
                    "type": "list",
                    "id": list_id_bulk
                }
            }
        }
    }
}

response = klaviyo.Profiles.bulk_subscribe_profiles(subscribe_profiles_request_body)

# Example 3: Unsubscribe a profile from email and SMS marketing
customer_email_unsubscribe = 'CUSTOMER_EMAIL'  # replace with the email address
customer_phone_number_unsubscribe = '+15555555555'  # replace with the phone number
list_id_unsubscribe = 'YOUR_LIST_ID'  # replace with your list ID

unsubscribe_profile_request_body = {
    "data": {
        "type": "profile-subscription-bulk-delete-job",
        "attributes": {
            "profiles": {
                "data": [
                    {
                        "type": "profile",
                        "attributes": {
                            "subscriptions": {
                                "email": {
                                    "marketing": {
                                        "consent": "UNSUBSCRIBED"
                                    }
                                },
                                "sms": {
                                    "marketing": {
                                        "consent": "UNSUBSCRIBED"
                                    },
                                    "transactional": {
                                        "consent": "UNSUBSCRIBED"
                                    }
                                }
                            },
                            "email": customer_email_unsubscribe,
                            "phone_number": customer_phone_number_unsubscribe
                        }
                    }
                ]
            },
            "historical_import": False
        },
        "relationships": {
            "list": {
                "data": {
                    "type": "list",
                    "id": list_id_unsubscribe
                }
            }
        }
    }
}

response = klaviyo.Profiles.bulk_unsubscribe_profiles(unsubscribe_profile_request_body)

File: profiles/subscribe_and_unsubscribe.js
// this sample JavaScript code subscribes and unsubscribes a profile from email and SMS marketing.
// see https://developers.klaviyo.com/en/reference/bulk_subscribe_profiles and https://developers.klaviyo.com/en/reference/bulk_unsubscribe_profiles for more details.

import { ApiKeySession, ProfilesApi } from 'klaviyo-api'

// load private API key as an environment variable. Note that this API key needs to have write access to profiles, lists, and subscriptions.
// if using a .env file, ensure to load it using dotenv or similar package. 
const apiKey = process.env.KLAVIYO_API_KEY
if (!apiKey) {
  throw new Error('Please set the KLAVIYO_API_KEY environment variable.')
}

const session = new ApiKeySession(apiKey)
const profilesApi = new ProfilesApi(session)

// example 1: subscribe a profile to email and SMS marketing.
const customerEmail =  'CUSTOMER_EMAIL' // replace with the email address of the profile you want to subscribe
const customerPhoneNumber = 'CUSTOMER_PHONE_NUMBER' // replace with the phone number of the
const listId = 'YOUR_LIST_ID' // replace with the list ID you want to subscribe the profile to

const subscribeProfileRequestBody = {
  data: {
    type: "profile-subscription-bulk-create-job",
    attributes: {
      profiles: {
        data: [
          {
            type: "profile",
            attributes: {
              subscriptions: {
                email: {
                  marketing: {
                    consent: "SUBSCRIBED"
                  }
                },
                sms: {
                  marketing: {
                    consent: "SUBSCRIBED"
                  },
                  transactional: {
                    consent: "SUBSCRIBED"
                  }
                }
              },
              email: customerEmail,
              phone_number: customerPhoneNumber
            }
          }
        ]
      },
      historical_import: false
    },
    relationships: {
      list: {
        data: {
          type: "list",
          id: listId
        }
      }
    }
  }
}

profilesApi.bulkSubscribeProfiles(subscribeProfileRequestBody)

// example 2: subscribe one profile to email marketing, and one to SMS marketing and transactional SMS.
const customerEmail1 = 'CUSTOMER_EMAIL_1' // replace with the first customer's email
const customerEmail2 = 'CUSTOMER_EMAIL_2' // replace with the second customer's email
const customerPhone2 = 'CUSTOMER_PHONE_2' // replace with the second customer's phone number
const listIdBulk = 'YOUR_LIST_ID' // replace with the list ID you want to subscribe the profiles to

const subscribeProfilesRequestBody = {
  data: {
    type: "profile-subscription-bulk-create-job",
    attributes: {
      profiles: {
        data: [
          {
            type: "profile",
            attributes: {
              subscriptions: {
                email: {
                  marketing: {
                    consent: "SUBSCRIBED"
                  }
                }
              },
              email: customerEmail1
            }
          },
          {
            type: "profile",
            attributes: {
              subscriptions: {
                sms: {
                  marketing: {
                    consent: "SUBSCRIBED"
                  },
                  transactional: {
                    consent: "SUBSCRIBED"
                  }
                }
              },
              email: customerEmail2,
              phone_number: customerPhone2
            }
          }
        ]
      },
      historical_import: false
    },
    relationships: {
      list: {
        data: {
          type: "list",
          id: listIdBulk
        }
      }
    }
  }
}

profilesApi.bulkSubscribeProfiles(subscribeProfilesRequestBody)

// example 3: unsubscribe a profile from email and SMS marketing.
const customerEmailUnsubscribe =  'CUSTOMER_EMAIL' // replace with the email address of the profile you want to unsubscribe
const customerPhoneNumberUnsubscribe = 'CUSTOMER_PHONE_NUMBER' // replace with the phone number
const listIdUnsubscribe = 'YOUR_LIST_ID' // replace with the list ID you want to unsubscribe the profile from

const unsubscribeProfileRequestBody = {
  data: {
    type: "profile-subscription-bulk-delete-job",
    attributes: {
      profiles: {
        data: [
          {  
          type: "profile",
            attributes: {
              subscriptions: {
                email: {
                  marketing: {
                    consent: "UNSUBSCRIBED"
                  }
                },
                sms: {
                  marketing: {
                    consent: "UNSUBSCRIBED"
                  },
                  transactional: {
                    consent: "UNSUBSCRIBED"
                  }
                }
              },
              email: customerEmailUnsubscribe,
              phone_number: customerPhoneNumberUnsubscribe
            }
          }
        ]
      },
      historical_import: false
    },
    relationships: {
      list: {
        data: {
          type: "list",
          id: listIdUnsubscribe
        }
      }
    }
  }
} 

profilesApi.bulkUnsubscribeProfiles(unsubscribeProfileRequestBody)
